app/models/upload.rb

```diff
@@ -1,196 +1,351 @@
+                      type: 'Text',
+                      prompt: b['Text']['prompt']
+                    }
+                  end
+                elsif b.key? 'TrueFalse'
+                  {
+                    type: 'YesNo',
+                    yesLabel: 'True',
+                    noLabel: 'False',
+                    prompt:
+                      if b['TrueFalse'] == !!b['TrueFalse']
+                        []
+                      else
+                        b['TrueFalse']['prompt']
+                      end
+                  }
+                elsif b.key? 'YesNo'
+                  {
+                    type: 'YesNo',
+                    prompt:
+                      if b['YesNo'] == !!b['YesNo']
+                        []
+                      else
+                        b['YesNo']['prompt']
+                      end
+                  }
+                else
+                  raise 'Bad question type.'
+                end
+              else
+                raise 'Bad body item.'
+              end
+            end
+          }.compact
+        end
+      }.compact
+    end
+    @info =
+      {
+        policies: properties['policies'] || [],
+        contents: {
+          questions: questions,
+          reference: e_reference,
+          instructions: exam_info['instructions']
+        }.compact,
+        answers: answers
+    }
   end
 
-  def extract_contents!(mimetype)
-    return if Dir.exist?(extracted_path)
-
-    extract_contents_to!(mimetype, extracted_path, postprocess: true, force_readable: true)
+  def rec_path(base_path, path)
+    if path.symlink?
+      {
+        path: path.basename.to_s,
+        link_to: path.dirname.join(File.readlink(path)),
+        broken: (!File.exist?(File.realpath(path)) rescue true)
+      }
+    elsif path.file?
+      {
+        path: path.basename.to_s,
+        full_path: path,
+        relPath: path.relative_path_from(base_path)
+      }
+    elsif path.directory?
+      {
+        path: path.basename.to_s,
+        relPath: path.relative_path_from(base_path),
+        children: path.children.sort.collect do |child|
+          rec_path(base_path, child)
+        end
+      }
+    end
   end
 
-  def extract_contents_to!(mimetype, extracted_path, postprocess: false, force_readable: false)
-    extracted_path.mkpath
-    ArchiveUtils.extract(original_path.to_s, mimetype, extracted_path.to_s, force_readable: force_readable)
-    return unless postprocess
+  def extract_contents!
+    mimetype = @upload.content_type
+    # TODO rewrite archiveutils to take callbacks for reading and writing files
+    ArchiveUtils.extract(@upload.path.to_s, mimetype, @dir, force_readable: true)
 
     found_any = false
-    Find.find(extracted_path) do |f|
+    Find.find(@dir) do |f|
       next unless File.file? f
 
       found_any = true
       next if File.extname(f).empty?
 
-      Postprocessor.process(extracted_path, f)
+      Postprocessor.process(@dir, f)
```

Comment:
#81

Reply:
I see the issue; I'd still like to see a test-case exam that uses the preprocessor, so that we know whether our jsonb handling of files is working properly.